pyat/at/lattice/lattice_object.py

```diff
@@ -90,13 +87,7 @@ def full_scan(elems, keep_all=False, **kwargs):
                 attributes['name'] = ''
                 if 'energy' not in kwargs:
                     # Guess energy from the Energy attribute of the elems
-                    if not energies:
-                        raise AtError('Lattice energy is not defined')
-                    energy = max(energies)
-                    if min(energies) < energy:
-                        warn(AtWarning('Inconsistent energy values, '
-                                       '"energy" set to {0}'.format(energy)))
-                    attributes['energy'] = energy
+                    attributes['energy'] = get_ring_energy(elems)
```

Comment:
There is a problem here:
I made significant modifications to `lattice_object` in the `plot_preparation` branch. To avoid unnecessary `list` copies (build a `list` first, and then give it to `super(Lattice, self).__init__`,  the `list` constructor, which will itself build a copy of this `list`), `full_scan` now accepts any iterable as input and returns an iterator over the elements. So it is executed **within** `super(Lattice, self).__init__`. The gain is most effective when the input to `full_scan` is itself an iterator. In such a case, after the first loop over elements, `get_ring_energy` will get an empty iterator, and fail.

Example: we have a Lattice `ring`, without `RingParam` element but with an `RFCavity` giving the energy. Let's assume we want to filter this lattice with some condition:

`ring2=Lattice([elem for elem in ring if <some condition>])`
Here we build a list first, which will be copied by the list constructor. `get_ring_energy` will work.

`ring2=Lattice(elem for elem in ring if <some condition>)`
Here we have no intermediate list, `<some condition>` will be tested within the list constructor and `get_ring_energy` will fail.

I think we should keep the possibility of accepting any iterable (sequence or iterator) as input, not only a sequence.

For the time being, I suggest to keep lattice_object unchanged to avoid merging problems. Keep all your other modifications as they are, let's merge both branches and start over then with this code factorisation.

Reply:
Ok, I will revert the modifications to `lattice_object` for now. Yes, I am learning that working on multiple branches at once is a recipe for problems, the test on #102 is failing because it relies on changes made here.